### PR TITLE
Support placement preferences in stack deployment

### LINKED ISSUE #32584: Support placement preferences in stack deploy
#30725 added a `--placement-pref` CLI option, but I don't believe it's supported within compose files yet.

cc @dnephin

### PR DESCRIPTION
**Pull request by denverdino**
Fix #32584 Support placement preferences in stack deployment

Signed-off-by: Li Yi <denverdino@gmail.com>

**- What I did**
Following the compose definition in https://github.com/docker/swarmkit/pull/1512 , I support the placement preferences in docker stack deploy

**- How I did it**
Add the placement preferences in docker compose scheme

**- How to verify it**

Create a docker swarm with 3 nodes, and then add the "az" label to the nodes

docker node update --label-add az=az1 node1
docker node update --label-add az=az1 node2
docker node update --label-add az=az2 node3

Create docker-compose.yaml as following

```
version: "3.3"
services:
  nginx:
    image: nginx
    deploy:
      mode: replicated
      replicas: 2
      placement:
        preferences:
          - spread: node.labels.az
```

docker stack deploy -c docker-compose.yaml test

The nginx container will be spread on the nodes with different az

**- A picture of a cute animal (not mandatory but encouraged)**

Image of Panda

### CODE REVIEW

File: cli/compose/schema/data/config_schema_v3.3.json [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+            "window": {"type": "string", "format": "duration"}
+          },
+          "additionalProperties": false
+        },
+        "placement": {
+          "type": "object",
+          "properties": {
+            "constraints": {"type": "array", "items": {"type": "string"}},
+			"preferences": {
+			  "type": "array",
```
Comment (dnephin) [inline cli/compose/schema/data/config_schema_v3.3.json]: The formatting is off here, are you using tabs? This file uses spaces

File: cli/compose/schema/schema_test.go [hunk 2]
```diff
@@ -50,3 +50,27 @@ func TestValidateInvalidVersion(t *testing.T) {
...
+								"spread": "node.labels.aliyun.zone",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	assert.NoError(t, Validate(config, "3.3"))
```
Comment (dnephin) [inline cli/compose/schema/schema_test.go]: This test would be more appropriate as a test in `cli/compose/loader/`. There is a TestFull that you can add to

The tests in this package are for the schema loading and validation.

**Review by dnephin — COMMENTED** [review]
The new version of the schema is missing the consistency field from 3.2.

File: cli/compose/schema/schema_test.go (hunk 2 above)
Comment (denverdino) [inline cli/compose/schema/schema_test.go]: Test case added. Thanks for comments

**Comment by dnephin** [comment]
Design LGTM

File: cli/compose/loader/full-example.yml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-version: "3.2"
+version: "3.3"
```
Comment (vdemeester) [inline cli/compose/loader/full-example.yml]: cc @dnephin

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

File: cli/compose/loader/full-example.yml (hunk 3 above)
Comment (dnephin) [inline cli/compose/loader/full-example.yml]: Ya this is necessary as we expand full example to include the new fields

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by vdemeester** [comment]
!rebuild powerpc #33041

**Comment by vdemeester** [comment]
Moving to docs-review, cc @thaJeztah

**Comment by ColinHebert** [comment]
FYI, due to moby/moby#32694, this PR should be closed and opened again in https://github.com/docker/cli

It will require the changes in docker/cli#32 to be ported as is.

**Comment by denverdino** [comment]
thx for notification

**Comment by denverdino** [comment]
Open a new PR docker/cli#35

**Pull request closed without merging**